pkg/KubeArmorOperator/internal/controller/resources.go

```diff
@@ -0,0 +1,585 @@
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	deployments "github.com/kubearmor/KubeArmor/deployments/get"
+	crds "github.com/kubearmor/KubeArmor/pkg/KubeArmorController/crd"
+	"github.com/kubearmor/KubeArmor/pkg/KubeArmorOperator/common"
+	v1 "k8s.io/api/admissionregistration/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func generateDaemonset(name, enforcer, runtime, socket, runtimeStorage string) *appsv1.DaemonSet {
+	enforcerVolumes, enforcerVolumeMounts := genEnforcerVolumes(enforcer)
+	runtimeVolumes, runtimeVolumeMounts := genRuntimeVolumes(runtime, socket, runtimeStorage)
+	vols := []corev1.Volume{}
+	volMnts := []corev1.VolumeMount{}
+	vols = append(vols, enforcerVolumes...)
+	vols = append(vols, runtimeVolumes...)
+	vols = append(vols, common.CommonVolumes...)
+	volMnts = append(volMnts, enforcerVolumeMounts...)
+	volMnts = append(volMnts, runtimeVolumeMounts...)
+	volMnts = append(volMnts, common.CommonVolumesMount...)
+
+	daemonset := deployments.GenerateDaemonSet("generic", common.Namespace)
+	daemonset.Name = name
+	labels := map[string]string{
+		common.EnforcerLabel:       enforcer,
+		common.RuntimeLabel:        runtime,
+		common.RuntimeStorageLabel: runtimeStorage,
+		common.SocketLabel:         socket,
+		common.OsLabel:             "linux",
+	}
+	daemonset.Spec.Template.Spec.NodeSelector = labels
+	daemonset.Spec.Template.Labels = labels
+	daemonset.Spec.Template.Spec.ServiceAccountName = "kubearmor"
+	daemonset.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: labels,
+	}
+	if deployment_uuid != "" {
+		daemonset.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: "apps/v1",
+				Kind:       "Deployment",
+				Name:       deployment_name,
+				UID:        deployment_uuid,
+			},
+		}
+	}
+	daemonset.Spec.Template.Spec.Volumes = vols
+	daemonset.Spec.Template.Spec.InitContainers[0].VolumeMounts = common.CommonVolumesMount
+	daemonset.Spec.Template.Spec.Containers[0].VolumeMounts = volMnts
+	daemonset.Spec.Template.Spec.Containers[0].Args = append(daemonset.Spec.Template.Spec.Containers[0].Args, "-criSocket=unix:///"+strings.ReplaceAll(socket, "_", "/"))
+	daemonset = addOwnership(daemonset).(*appsv1.DaemonSet)
+	fmt.Printf("generated daemonset: %v", daemonset)
+	return daemonset
+}
+
+func genEnforcerVolumes(enforcer string) (vol []corev1.Volume, volMnt []corev1.VolumeMount) {
+	if enforcer == "none" {
+		return nil, nil
+	}
+	for _, e := range strings.Split(enforcer, ".") {
+		vol = append(vol, common.EnforcerVolumes[e]...)
+		volMnt = append(volMnt, common.EnforcerVolumesMounts[e]...)
+	}
+	return
+}
+
+func genRuntimeVolumes(runtime, runtimeSocket, runtimeStorage string) (vol []corev1.Volume, volMnt []corev1.VolumeMount) {
```

Comment:
How do we deal with AWS BottleRocket where dockershim is being used?

Ref https://github.com/kubearmor/KubeArmor/blob/0dd8f8bb9507aba1a5361092d85372f0fffccd34/deployments/BottleRocket/kubearmor.yaml#L2652

Reply:
i think we missed this case, need to handle this. thanks for pointing this out.

Reply:
it has been fixed